perf: transmit RASET/CASET only when changed

ardnew opened the pull request:
These display drivers always write color data to the screen region defined by their row address (`RA`) and column address (`CA`) registers. 

The SPI command to update the addresses in these registers are `RASET` and `CASET`. It is not necessary to re-transmit `RASET`/`CASET` when refreshing the color data in the same address window repeatedly. 

Apparently, it is **extremely** common for graphics drivers to refresh a fixed region of the display (full-screen refresh, dirty rectangles, bitmap fonts/sprites, etc.). So this optimization yields a substantial performance improvement in nearly all cases without penalty.

Also, the call to `writecommand(ST7735_RAMWR)` has been refactored into the `setAddr` routine. You can easily see from the patch that this command is always transmitted immediately following calls to `setAddr`. So this merely reduces a lot of duplicate lines.

ardnew committed:
perf: transmit RASET/CASET only when changed

refactor: move RAMWR command into setAddr

h4yn0nnym0u5e commented on src/ST7735_t3.h:
```diff
@@ -234,12 +234,35 @@ class ST7735_t3 : public Print
-        writedata16(y1+_ystart);   // YEND
+        // Keep a copy of the last address window transmitted.
+	// Do not transmit it again until the window actually changes.
+        static uint16_t _x0 = 0xFFFF, _y0 = 0xFFFF, 
```
This doesn't work if multiple instances of the display are connected - you need to make the stored values private class members.

ardnew replied:
Replaced static local vars with protected instance members

h4yn0nnym0u5e commented on src/ST7735_t3.h:
```diff
@@ -234,12 +234,35 @@ class ST7735_t3 : public Print
+        static uint16_t _x0 = 0xFFFF, _y0 = 0xFFFF, 
+                        _x1 = 0xFFFF, _y1 = 0xFFFF;
+
+        // TBD: Guard these under the same conditions as CASET/RASET below?
```
No, because the start might have been changed, and then this gets called with the same screen co-ordinates. You _have_ to do the calculation ... but as you say, that's not the bottleneck.

h4yn0nnym0u5e replied:
Turns out ... no. The start is set at initialisation, and never changes.

ardnew replied:
Are you saying this is just an unnecessary comment that can be removed, or suggesting a different change?

h4yn0nnym0u5e replied:
TLDR: no need to guard here.

I made the comment on the erroneous assumption that `_xstart` and `_ystart` were values that might change as the program runs. But on deeper investigation they're set only once, during initialisation, to cope with displays that have the hardware wired to start at non-zero row or column. So they shouldn't need guarding / checking as part of the determination of whether CASET / RASET is needed.

All that being said, and after I decided to adopt (most of) your approach, I got this feedback that some displays behave badly if you don't do both: https://forum.pjrc.com/index.php?threads/st7796-teensyduino-support.76510/post-359786

It's rather vague, unfortunately, so I'm going to leave it in and see if I get any adverse events; so far all my testing has been with an ST7796 display, but I now have ST7735 and ST7789 displays to try.

ardnew committed:
fix: replace static locals with instance vars

Per review comment, static local vars will cause conflict when using multiple
displays. Each driver instance needs to maintain its own copy across calls to
setAddr.
